src/Illuminate/Cache/RedisLock.php

```diff
@@ -49,6 +49,18 @@ public function acquire()
      */
     public function release()
     {
-        $this->redis->del($this->name);
+        if ($this->canRelease()) {
+            $this->redis->del($this->name);
```

Comment:
There is a race condition here.  It's possible for `canRelease` to return true, the lock to then be released because it was expired, and another client to acquire the lock before `del` is called.

With Redis you can use a Lua script to perform an atomic operation.  Since scripts are blocking no other client can acquire the lock between the check and the delete.

```lua
if redis.call("get",KEYS[1]) == ARGV[1] then
    return redis.call("del",KEYS[1])
else
    return 0
end
```

Unfortunately I don't think there is a good solution for memcached.

https://redis.io/topics/distlock

Reply:
Couldn't we use a transaction for that? The Redis docs say:
> It can never happen that a request issued by another client is served in the middle of the execution of a Redis transaction.

For Memcached the `cas` command can be used setting it to `null`. Though this would eliminate the possibility of storing the lock for later release as @taylorotwell pointed out above as the `cas` assumes a state on the client: http://php.net/manual/de/memcached.cas.php

Reply:
> Couldn't we use a transaction for that? 

You need to check if the key matches the token before deleting it.  I don't think there is an equivalent Redis command that does that.  Redis transactions aren't like database transactions, the GET will just be queued until EXEC and you can't read the value of the key in PHP before you call `EXEC`.